File: sdk/typescript/examples/11-bluetooth.ts
// 11 — Bluetooth (Linux only)
//
// List paired devices, optionally scan for new ones, and connect/disconnect by
// MAC address. Works on Linux hosts where rockboxd has access to BlueZ.
//
//   bun run examples/11-bluetooth.ts                    # list paired devices
//   bun run examples/11-bluetooth.ts scan               # scan 10s for devices
//   bun run examples/11-bluetooth.ts connect AA:BB:..   # connect by address
//   bun run examples/11-bluetooth.ts disconnect AA:BB:..

import type { BluetoothDevice } from '../src/index.js';
import { createClient } from './_client.js';

function format(d: BluetoothDevice): string {
  const flags = [
    d.connected ? 'connected' : '',
    d.paired    ? 'paired'    : '',
    d.trusted   ? 'trusted'   : '',
  ].filter(Boolean).join(', ');
  const rssi = d.rssi != null ? ` ${d.rssi} dBm` : '';
  return `  ${d.address}  ${d.name.padEnd(28)}${rssi}  [${flags}]`;
}

const client = createClient();
const cmd = process.argv[2] ?? 'list';

try {
  if (cmd === 'scan') {
    console.log('Scanning for 10s...');
    const found = await client.bluetooth.scan(10);
    console.log(`Found ${found.length} devices:`);
    found.forEach((d) => console.log(format(d)));
  } else if (cmd === 'connect' && process.argv[3]) {
    await client.bluetooth.connect(process.argv[3]);
    console.log(`Connected to ${process.argv[3]}`);
  } else if (cmd === 'disconnect' && process.argv[3]) {
    await client.bluetooth.disconnect(process.argv[3]);
    console.log(`Disconnected from ${process.argv[3]}`);
  } else {
    const devices = await client.bluetooth.devices();
    console.log(`Paired devices (${devices.length}):`);
    devices.forEach((d) => console.log(format(d)));
  }
} catch (err) {
  // Non-Linux hosts return "Bluetooth is only supported on Linux".
  console.error('Bluetooth call failed:', err instanceof Error ? err.message : err);
  process.exit(1);
}

File: sdk/elixir/examples/11_bluetooth.exs
# 11 — Bluetooth (Linux only)
#
#   mix run examples/11_bluetooth.exs                    # list paired devices
#   mix run examples/11_bluetooth.exs scan               # scan 10s for devices
#   mix run examples/11_bluetooth.exs connect AA:BB:..   # connect by address
#   mix run examples/11_bluetooth.exs disconnect AA:BB:..

Code.require_file("_helper.exs", __DIR__)

client = Examples.Helper.client()

format = fn d ->
  flags =
    [
      if(d.connected, do: "connected"),
      if(d.paired, do: "paired"),
      if(d.trusted, do: "trusted")
    ]
    |> Enum.reject(&is_nil/1)
    |> Enum.join(", ")

  rssi = if d.rssi, do: " #{d.rssi} dBm", else: ""
  "  #{d.address}  #{String.pad_trailing(d.name || "", 28)}#{rssi}  [#{flags}]"
end

case System.argv() do
  ["scan"] ->
    IO.puts("Scanning for 10s...")
    {:ok, found} = Rockbox.Bluetooth.scan(client, 10)
    IO.puts("Found #{length(found)} devices:")
    for d <- found, do: IO.puts(format.(d))

  ["connect", addr] ->
    :ok = Rockbox.Bluetooth.connect(client, addr)
    IO.puts("Connected to #{addr}")

  ["disconnect", addr] ->
    :ok = Rockbox.Bluetooth.disconnect(client, addr)
    IO.puts("Disconnected from #{addr}")

  _ ->
    {:ok, devices} = Rockbox.Bluetooth.devices(client)
    IO.puts("Paired devices (#{length(devices)}):")
    for d <- devices, do: IO.puts(format.(d))
end
